GGRC-4311 Implement GDrive interaction code on BE

olegnatsevsky opened the pull request:
# Issue description
Implement GDrive interaction code on BE.

# Steps to test changes
- attach evidence file to Audit, Assessment, Control and check that all work as before.
- check import/export flow.
- integration and unit tests.

# Solution description
Move gdrive API manipulation to the BE.
Reimplement file name postfixes.
Run ‘gdrive file copy’ flow in GGRC Document obj creation.

# Sanity checklist

- [x] I have clicked through the app to make sure my changes work and not break the app.
- [x] I have applied the correct milestone and labels.
- [x] My changes fix the issue described in the description (and do nothing else). 🤞
- [x] My changes are covered by tests.
- [x] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [x] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [x] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

<!-- If your PR includes a migration include the additional checklist items
# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)
-->

# PR Review checklist

- [x] The changes fix the issue and don't cause any apparent regressions.
- [x] Labels and milestone are correctly set.
- [x] The solution description matches the changes in the code.
- [x] There is no apparent way to improve the performance & design of the new code.
- [x] The pull request is opened against the correct base branch.
- [x] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

<!-- If your code is not finished yet can include a TODO check list
# TODO

- [ ] First item on the TODO
-->

valchanin commented:
<img width="735" alt="screen shot 2018-02-21 at 11 32 24" src="https://user-images.githubusercontent.com/12893937/36470136-eed4a85c-16fa-11e8-8033-2b4f67b7ce67.png">

valchanin commented:
@olegnatsevsky you cherry-picked wrong fe commits

NastassiaKiseleva commented:
Adding needs work because of failed integration tests and code style

tor4 commented:
Looks like I found a small issue.
When I upload evidence to assessment from my laptop, there are 2 files in folder: original one and with postfix:
![image](https://user-images.githubusercontent.com/3773902/36778517-fa60fc38-1c7d-11e8-926f-6b6251774e4f.png)

GDrive:
![image](https://user-images.githubusercontent.com/3773902/36778541-0f992ddc-1c7e-11e8-95a1-28d3ddda1a2f.png)

tor4 commented:
Also sometimes BE returns Internal Server Error response. Unfortunately I cannot provide exact steps to reproduce it. 
Sometimes it happens when I clear cookies to revoke BE gdrive authorization. Attach or remove folder. And then try to upload new evidences.
Also FE does not notify user that error happened.

![image](https://user-images.githubusercontent.com/3773902/36781312-96179968-1c86-11e8-88a4-b0797976d7ec.png)

tor4 commented:
When I try to attach file to Risk Assessment there's a 400 BAD REQUEST response ("Allowed types are: Audit, Control, Assessment, Issue.").
Risk Assessment should be supported on BE or 'Attach' button should be removed from FE.

tor4 commented on src/ggrc-client/js/components/gdrive/gdrive_picker_launcher.js:
```diff
@@ -166,23 +172,28 @@ import {
+          });
+
+          return backendGdriveClient.withAuth(()=> {
+            return CMS.Models.Document.create(modelData, {global: false});
```
Please add a comment why `global` is `false`.

olegnatsevsky commented on src/ggrc-client/js/components/gdrive/gdrive_picker_launcher.js:
```diff
@@ -296,8 +83,7 @@ import RefreshQueue from '../../models/refresh_queue';
 
-          this.addFilesSuffixes({}, files)
-            .then(this.handle_file_upload.bind(this))
+          this.createDocumentModel(files)
```
It seems BE not handles properly  'validation errors' (Error message not in JSON format). This generic handler that should be updated is separate scope. As I've added Risk Assessment support, this error message is not shown to users anymore.

olegnatsevsky commented:
@tor4 Thanks for the review!
1) File duplication: fixed.
2) Server Error response: related to SSL: CERTIFICATE_VERIFY_FAILED error during request to "https://www.googleapis.com/discovery/v1/apis/drive/v3/rest" It weird, but after recreation of the container, I can't reproduce it anymore. We need to double check it at QA platform.
3) Risk Assessment: fixed.
4) Commented.

olegnatsevsky commented on src/ggrc/gdrive/__init__.py:
```diff
@@ -68,38 +70,47 @@ def is_gdrive_authorized():
+  """Second step of the OAuth2"""
+  if 'code' not in flask.request.args:
+    raise Unauthorized('Broken OAuth2 flow, go to /auth_gdrive first')
+  # Cross Site Request Forgery (XRSF) guard.
```
Fixed.

olegnatsevsky commented on src/ggrc/gdrive/file_actions.py:
```diff
@@ -47,12 +32,36 @@ def hande_http_error(ex):
+    result = drive_service.files().create(body=file_metadata,
+                                          media_body=media,
+                                          fields='id, name, parents').execute()
+    return result
```
Done.

olegnatsevsky commented on src/ggrc/gdrive/__init__.py:
```diff
@@ -68,38 +70,47 @@ def is_gdrive_authorized():
     raise Unauthorized('')
 
 
-@app.route("/authorize")
```
Ok, will separate next PR.

olegnatsevsky commented on src/ggrc/gdrive/file_actions.py:
```diff
@@ -68,19 +77,58 @@ def get_gdrive_file(file_data):
+  """Helper for sanitize filename"""
+  original_name, extension = path.splitext(original_name)
+  # remove an old postfix
+  original_name = original_name.split('_ggrc')[0]
```
Good point. Fixed

olegnatsevsky commented on src/ggrc/gdrive/file_actions.py:
```diff
@@ -68,19 +77,58 @@ def get_gdrive_file(file_data):
+  original_name = original_name.split('_ggrc')[0]
+  new_name = ''.join([original_name, postfix]).strip('_')
+  # sanitaze file name
+  new_name = ''.join([char if char.isalnum() or char in ALLOWED_FILENAME_CHARS
```
Good point, I've added this strings to the unittest.

olegnatsevsky commented on src/ggrc/models/document.py:
```diff
@@ -9,17 +9,21 @@
-class Document(Roleable, Relatable, Base, mixins.Titled, Indexed, db.Model):
-  """Audit model."""
+class Document(Roleable, Relatable, Base, mixins.Titled, Indexed, db.Model,
+               bfh.BeforeFlushHandleable):
```
Ok

olegnatsevsky commented on src/ggrc/gdrive/file_actions.py:
```diff
@@ -68,19 +77,58 @@ def get_gdrive_file(file_data):
+  original_name, extension = path.splitext(original_name)
+  # remove an old postfix
+  original_name = original_name.split('_ggrc')[0]
+  new_name = ''.join([original_name, postfix]).strip('_')
```
You are right. We don't need to strip here. Fixed.

tor4 commented:
@olegnatsevsky @valchanin  Please fix tests.

tor4 commented:
If I upload files and BE is not authorized, only one document is created and saving spinner does not disappear.
![image](https://user-images.githubusercontent.com/3773902/36933340-4737fdc4-1ee8-11e8-9f16-bdd05a7f01e3.png)

NastassiaKiseleva commented:
@olegnatsevsky Oleg, please fix integration tests

alieh-rymasheuski commented on test/unit/ggrc/models/test_file_actions.py:
```diff
@@ -1,3 +1,4 @@
+# coding=utf-8
```
Please use the same format as in other places in the project:

```python
# -*- coding: utf-8 -*-

```

olegnatsevsky replied:
Done

alieh-rymasheuski commented on test/unit/ggrc/models/test_file_actions.py:
```diff
@@ -17,7 +18,6 @@ class TestFileActions(unittest.TestCase):
       # original_file_name, postfix, expected
```
Please add a test case when there are special characters in the postfix itself (like, an assessment may have slug equal to `!@#%$&^*()`, and it will be still valid).

olegnatsevsky replied:
Done.

alieh-rymasheuski commented on src/ggrc/gdrive/__init__.py:
```diff
@@ -26,6 +26,10 @@
 logger = getLogger(__name__)
 
 
+class GdriveUnauthorized(Unauthorized):
```
You could also write a docstring that this exception is a "soft" exception as we always know how to recover from it.

alieh-rymasheuski commented on test/integration/ggrc/models/test_document.py:
```diff
@@ -2,23 +2,190 @@
+            'type': 'Control'
+        })
+    # pylint: disable=protected-access
+    self.assertEqual(control, document._get_documentable_obj())
```
This test case and below should be implemented through API calls, rather than factories with mocks. I'd expect it to be like:
```python
# positive test
response = POST("/api/documents", [{document_json, "documentable_obj"={"id": control.id, "type": control.type}}])
self.assert200(response)
self.assert_stuff_on_response_body(response)

# negative tests
response = POST("/api/documents", [{document_json, "documentable_obj"={broken_somehow}}])
self.assert400(response)
self.assertEqual(response.json["message"], your_cool_error_message)
```
